### PR TITLE
Added the possibility to zoom the map

### PR DESCRIPTION
**Pull request by Reizoukodesu**
Adds the possibility to zoom the map.

Because of my inactivity the previous PR was closed. This PR refs #9906
I fixed the remaining coding issues but didn't find a option to reopen the previous PR.
Hope that I'm not too late.

### CODE REVIEW

**Comment by GraionDilach** [comment]
OpenRA.Mods.Common/Widgets/EditorViewportControllerWidget.cs(78,48): error CS1061: Type `float[]' does not contain a definition for `Count' and no extension method `Count' of type `float[]' could be found. Are you missing `System.Linq' using directive?

OpenRA.Mods.Common/Widgets/EditorViewportControllerWidget.cs(81,25): error CS1061: Type `float[]' does not contain a definition for `ElementAt' and no extension method `ElementAt' of type `float[]' could be found. Are you missing `System.Linq' using directive?

**Comment by Sapena** [comment]
I had a little mess with my commits. They are now squashed.

File: OpenRA.Game/Graphics/Viewport.cs [hunk 1]
```diff
@@ -74,6 +82,23 @@ public float Zoom
 		public static int TicksSinceLastMove = 0;
 		public static int2 LastMousePos;
 
+		float ClosestTo(float[] collection, float target)
+		{
+			var closest_value = collection.First();
+			var subtract_result = Math.Abs(closest_value - target);
```
Comment (Mailaender) [inline OpenRA.Game/Graphics/Viewport.cs]: Can you change those to camelCase please? https://github.com/OpenRA/OpenRA/wiki/Coding-Standard

**Comment by Mailaender** [comment]
Works as promised. :white_check_mark: Maybe also enable this for replays and spectators?

**Comment by Sapena** [comment]
Added replays and spectators

**Comment by obrakmann** [comment]
```
Exception of type `System.NullReferenceException`: Object reference not set to an instance of an object
  at OpenRA.Mods.Common.Widgets.ViewportControllerWidget.IsZoomAllowed (Single zoom) [0x00032] in /home/oliver/devel/openra/git/OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs:194 
  at OpenRA.Mods.Common.Widgets.ViewportControllerWidget.Zoom (Int32 direction) [0x00061] in /home/oliver/devel/openra/git/OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs:212 
  at OpenRA.Mods.Common.Widgets.ViewportControllerWidget.HandleMouseInput (MouseInput mi) [0x0004b] in /home/oliver/devel/openra/git/OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs:223 
```

**Comment by Sapena** [comment]
Thanks for the nice bug report, I now see the problem but, I don't understand the mechanics fully yet. I thought that localPlayer would always be set if it's no replay. If this isn't the case, how can i detect if a player is a spectator in the ViewportControllerWidget HandleMouseInput context? Just detect if localPlayer == null?

**Comment by obrakmann** [comment]
`LocalPlayer` is null when the user is a spectator (ie. they chose a spectator slot in the lobby) or in replays. `player.Spectating` only gets set when players become observers after they have lost (because there is no player object where that flag could be set for regular spectators).

So the test you want is `LocalPlayer == null || LocalPlayer.Spectating`, as already mentioned above.

**Comment by GraionDilach** [comment]
`LocalPlayer` is also `null` when the Disable visibility options devcheat is used.

**Comment by Sapena** [comment]
Everything is fixed now, at least I hope so. Let me know if there's something more to do.

**Comment by obrakmann** [comment]
Looks good to me, :+1: 

Could you please squash those commits together so this can be merged?

**Comment by obrakmann** [comment]
Superseded by #10821.

**Pull request closed without merging**